### ISSUE TITLE
Question: Where to invoke gevent's patch_all() when using pytest

### ISSUE THREAD

**Issue by LefterisJP** [pytest-dev/pytest#6210]
If you are familiar with gevent you will know that at the very beginning of each program it needs to monkey patch all:

```python
import gevent.monkey
gevent.monkey.patch_all()
```

So far for pytest I have been doing that at the very top of `conftest.py` and had no problems.

After upgrading `gevent` and `pytest` this seems to no longer work and leads to problems that come from [patching too late](https://github.com/gevent/gevent/issues/1481#issue-523822269).

The workaround I did was to make a wrapper script for pytest.

```python
from gevent import monkey  # isort:skip # noqa
monkey.patch_all()  # isort:skip # noqa

import pytest
pytest.main()
```

and call that script instead of pytest when I want to run tests. So the question is if there is any other better solution? What is the best place to invoke gevent's `patch_all()` when running pytest?

**Comment by RonnyPfannschmidt** [comment]
this is probably due to pytest importing certain conftests later now,
a pytest plugin to sort out gevent support seems necessary to do this correctly

**Comment by asottile** [comment]
this is the import path fwiw:

```console
$ pytest t.py
Traceback (most recent call last):
  File "/tmp/venv38/bin/pytest", line 5, in <module>
    from pytest import main
  File "/tmp/venv38/lib/python3.8/site-packages/pytest.py", line 12, in <module>
    from _pytest.debugging import pytestPDB as __pytestPDB
  File "/tmp/venv38/lib/python3.8/site-packages/_pytest/debugging.py", line 6, in <module>
    from doctest import UnexpectedException
  File "/usr/lib/python3.8/doctest.py", line 104, in <module>
    import unittest
  File "/usr/lib/python3.8/unittest/__init__.py", line 60, in <module>
    from .async_case import IsolatedAsyncioTestCase
  File "/usr/lib/python3.8/unittest/async_case.py", line 1, in <module>
    import asyncio
  File "/usr/lib/python3.8/asyncio/__init__.py", line 8, in <module>
    from .base_events import *
  File "/usr/lib/python3.8/asyncio/base_events.py", line 34, in <module>
    import ssl
  File "/tmp/venv38/lib/python3.8/ssl.py", line 93, in <module>
    raise AssertionError('oops')
AssertionError: oops
```

**Comment by blueyed** [comment]
Fixed by 07f20ccab then likely.

**Comment by blueyed** [comment]
@LefterisJP 
Additional information about when this changed for you would be helpful, i.e. it would be good if you could git-bisect it.
The above mentioned change should not be relevant for recent versions of pytest at least (it was there already in eaa882f3d.

**Comment by asottile** [comment]
I'm working on a ~plugin that'll "fix" this -- I imagine fixing this in pytest and all plugins for all gevent patches to be an endless game of wackamole which'll make all the code worse just for gevent's sake, we also need this at lyft for our stuff as well

https://github.com/asottile/pytest-gevent/

for @LefterisJP's case, they are using `web3` which provides a pytest plugin (which imports a bunch of stuff breaking `gevent` if it happens too early) so they're kinda stuck

**Comment by asottile** [comment]
I've released `pytest-gevent==1.0.0`.  compare with `web3` and `git+https://github.com/pytest-dev/pytest@features` installed:

```python
import gevent.monkey  # technically not needed when using pytest-gevent
gevent.monkey.patch_all()  # technically not needed when using pytest-gevent
def test():
    pass
```

```console
$ pytest t.py
============================= test session starts ==============================
platform linux -- Python 3.8.0, pytest-5.2.5.dev4+gf24f20a46, py-1.8.0, pluggy-0.13.0
rootdir: /tmp
plugins: web3-5.3.0
collected 1 item                                                               

t.py .                                                                   [100%]

=============================== warnings summary ===============================
t.py:2
  /tmp/t.py:2: MonkeyPatchWarning: Monkey-patching ssl after ssl has already been imported may lead to errors, including RecursionError on Python 3.6. It may also silently lead to incorrect behaviour on Python 3.7. Please monkey-patch earlier. See https://github.com/gevent/gevent/issues/1016. Modules that had direct imports (NOT patched): ['urllib3.util.ssl_ (/tmp/venv38/lib/python3.8/site-packages/urllib3/util/ssl_.py)', 'urllib3.util (/tmp/venv38/lib/python3.8/site-packages/urllib3/util/__init__.py)']. 
    gevent.monkey.patch_all()

-- Docs: https://docs.pytest.org/en/latest/warnings.html
======================== 1 passed, 1 warnings in 0.06s =========================
```

```console
$ pytest-gevent t.py
============================= test session starts ==============================
platform linux -- Python 3.8.0, pytest-5.2.5.dev4+gf24f20a46, py-1.8.0, pluggy-0.13.0
rootdir: /tmp
plugins: web3-5.3.0
collected 1 item                                                               

t.py .                                                                   [100%]

============================== 1 passed in 0.00s ===============================
```

**Comment by LefterisJP** [comment]
Hey @blueyed and @asottile 

Thank you so much for the quick responses and analysis of the situation.

My initial assumption that it had to do something with upgrading python from 3.7 to 3.8 was false. It was due to me trying to fix a lot of things together while testing with python 3.8 that I made that assumption.

The last known configuration that worked for our project was a combination of the following 3 projects and version:

```
pytest==5.0.1
web3==4.0.0
gevent==1.3.6
```

Our latest configuration which experiences the reported problem and needs the wrapper script is:

```
pytest==5.2.4
web3==5.3.0
gevent=1.5a2
```

I tried to go gradually from the working to the non-working config and indeed the problem is `web3`. If I update both `pytest` and `gevent` the problem is not occurring. But it starts to happen once `web3>=5.0.0`
